src/olm/account/mod.rs

```diff
@@ -439,6 +453,85 @@ impl Account {
 
         pickle.try_into()
     }
+
+    /// Create a dehydrated device from the account.
+    ///
+    /// A dehydrated device is a device that is stored encrypted on the server
+    /// that can receive messages when the user has no other active devices.
+    /// Upon login, the user can rehydrate the device (using
+    /// [`from_dehydrated_device`]) and decrypt the messages sent to the
+    /// dehydrated device.
+    ///
+    /// The account must be a newly-created account that does not have any Olm
+    /// sessions, since the dehydrated device format does not store sessions.
+    ///
+    /// Returns the ciphertext and nonce.  `key` is a 256-bit (32-byte) key for
+    /// encrypting the device.
+    ///
+    /// The format used here is defined in
+    /// [MSC3814](https://github.com/matrix-org/matrix-spec-proposals/pull/3814).
+    pub fn to_dehydrated_device(
+        &self,
+        key: &[u8; 32],
+    ) -> Result<DehydratedDeviceResult, crate::DehydratedDeviceError> {
+        use matrix_pickle::Encode;
+
+        use self::dehydrated_device::Pickle;
+        use crate::utilities::base64_encode;
+
+        let pickle: Pickle = self.into();
+        let mut encoded = pickle.encode_to_vec()?;
+        let cipher = ChaCha20Poly1305::new(key.into());
+        let rng = thread_rng();
+        let nonce = ChaCha20Poly1305::generate_nonce(rng);
+        let ciphertext = cipher.encrypt(&nonce, encoded.as_slice());
+        encoded.zeroize();
+        let ciphertext = ciphertext?;
+
+        Ok(DehydratedDeviceResult {
```

Comment:
Hmm, I know I said a struct would be nice, but perhaps we should encode it using `MatrixPickle` and then base64 encode it, though that requires again a version field.

We would have an inner pickle and an outer pickle, the inner pickle is the unencrypted account and the outer pickle is the ciphertext + nonce.

It's quite similar to the Olm/Megolm message packages, though there we use protobuf, so we could even pick that for consistency. What do you think about that?

Reply:
Yeah, I had considered that.  I prefer to have the data format be a standard format as much as possible, rather than increasing the number of custom binary formats that we have.  So I figured it would be better to return the nonce and ciphertext separately, and allow the SDK layer to put them in the JSON.